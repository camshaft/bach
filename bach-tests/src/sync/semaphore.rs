use crate::testing::Log;

use super::*;

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
enum Event {
    Start,
    SemaphoreAcquired { task: usize, permits: u32 },
    SemaphoreReleased { task: usize, permits: u32 },
}

impl crate::testing::Event for Event {
    fn is_start(&self) -> bool {
        matches!(self, Event::Start)
    }
}

#[test]
fn semaphore_permits() {
    bolero::check!().exhaustive().run(sim(|| {
        async {
            let semaphore = Semaphore::new(5);

            // Check initial permits
            assert_eq!(semaphore.available_permits(), 5);

            // Acquire some permits
            let permit1 = semaphore.acquire().await.unwrap();
            assert_eq!(semaphore.available_permits(), 4);

            // Add more permits
            semaphore.add_permits(3);
            assert_eq!(semaphore.available_permits(), 7);

            // Release the permit and check again
            drop(permit1);
            assert_eq!(semaphore.available_permits(), 8);
        }
        .primary()
        .spawn();
    }));
}

#[test]
fn semaphore_try_acquire() {
    bolero::check!().exhaustive().run(sim(|| {
        async {
            let semaphore = Semaphore::new(2);

            // Successfully try_acquire
            let permit1 = semaphore.try_acquire().unwrap();
            let permit2 = semaphore.try_acquire().unwrap();

            // This should fail
            assert!(semaphore.try_acquire().is_err());

            // Release one permit
            drop(permit1);

            // Now try_acquire should succeed again
            let _permit3 = semaphore.try_acquire().unwrap();

            // Clean up
            drop(permit2);
        }
        .primary()
        .spawn();
    }));
}

#[test]
fn semaphore_acquire_many() {
    bolero::check!().exhaustive().run(sim(|| {
        async {
            let semaphore = Arc::new(Semaphore::new(5));

            // Acquire multiple permits at once
            let permit = semaphore.acquire_many(3).await.unwrap();
            assert_eq!(semaphore.available_permits(), 2);

            // In a separate task to avoid deadlock
            {
                let semaphore = semaphore.clone();
                async move {
                    // Try to acquire too many
                    let too_many = semaphore.acquire_many(3).await.unwrap();

                    // Wait a bit
                    bach::time::sleep(std::time::Duration::from_millis(10)).await;

                    // Release the original permit
                    drop(too_many);
                }
                .primary()
                .spawn();
            }

            10.ms().sleep().await;
            drop(permit);
        }
        .primary()
        .spawn();
    }));
}

#[test]
fn semaphore_try_acquire_many() {
    bolero::check!().exhaustive().run(sim(|| {
        async {
            let semaphore = Semaphore::new(5);

            // Successfully try to acquire multiple permits
            let permit = semaphore.try_acquire_many(3).unwrap();
            assert_eq!(semaphore.available_permits(), 2);

            // This should fail
            assert!(semaphore.try_acquire_many(3).is_err());

            // This should succeed
            let _permit2 = semaphore.try_acquire_many(2).unwrap();
            assert_eq!(semaphore.available_permits(), 0);

            // Release first permit
            drop(permit);
            assert_eq!(semaphore.available_permits(), 3);
        }
        .primary()
        .spawn();
    }));
}

#[test]
fn semaphore_close() {
    bolero::check!().exhaustive().run(sim(|| {
        async {
            let semaphore = Semaphore::new(5);

            // Close the semaphore
            semaphore.close();

            // Further acquisitions should fail
            assert!(semaphore.try_acquire().is_err());

            // Async acquisition should also fail
            match semaphore.acquire().await {
                Err(_) => {} // Expected
                Ok(_) => panic!("Expected acquisition to fail after close"),
            };
        }
        .primary()
        .spawn();
    }));
}

#[test]
fn semaphore_interleavings_snapshot() {
    static LOG: Log<Event> = Log::new();

    bolero::check!().exhaustive().run(sim(|| {
        // Start a new event sequence
        LOG.push(Event::Start);

        // Create a semaphore with 2 permits
        let semaphore = Arc::new(Semaphore::new(2));

        // Launch multiple tasks that will contend for the semaphore
        for task_id in 0..4 {
            let semaphore = semaphore.clone();
            let permits = if task_id % 2 == 0 { 1u32 } else { 2u32 };

            async move {
                // Acquire permits
                if permits == 1 {
                    let _ = semaphore.acquire().await.unwrap();

                    LOG.push(Event::SemaphoreAcquired {
                        task: task_id,
                        permits,
                    });

                    // Do some work with the permit
                    bach::time::sleep(std::time::Duration::from_millis(5)).await;

                    // Release the permit (implicit when dropped)
                    LOG.push(Event::SemaphoreReleased {
                        task: task_id,
                        permits,
                    });
                } else {
                    let _ = semaphore.acquire_many(2).await.unwrap();

                    LOG.push(Event::SemaphoreAcquired {
                        task: task_id,
                        permits,
                    });

                    // Do some work with the permit
                    bach::time::sleep(std::time::Duration::from_millis(5)).await;

                    // Release the permit (implicit when dropped)
                    LOG.push(Event::SemaphoreReleased {
                        task: task_id,
                        permits,
                    });
                }
            }
            .primary()
            .spawn_named(format!("Task {task_id}"));
        }
    }));

    insta::assert_debug_snapshot!(LOG.check());
}

#[test]
#[should_panic = "Runtime stalled"]
fn semaphore_deadlock_detection() {
    sim(|| {
        // Create semaphores - one with limited permits
        let semaphore1 = Arc::new(Semaphore::new(1)); // Only 1 permit available
        let semaphore2 = Arc::new(Semaphore::new(1)); // Only 1 permit available

        // Task 1: acquire semaphore1, then try to acquire semaphore2
        {
            let semaphore1 = semaphore1.clone();
            let semaphore2 = semaphore2.clone();
            async move {
                // Acquire the first semaphore
                let permit1 = semaphore1.acquire().await.unwrap();

                // Try to acquire the second semaphore - this can deadlock
                // if task 2 is already holding semaphore2
                let permit2 = semaphore2.acquire().await.unwrap();

                // Release both semaphores
                drop(permit2);
                drop(permit1);
            }
            .primary()
            .spawn_named("Task 1");
        }

        // Task 2: acquire semaphore2, then try to acquire semaphore1
        {
            let semaphore1 = semaphore1.clone();
            let semaphore2 = semaphore2.clone();
            async move {
                // Acquire the second semaphore
                let permit2 = semaphore2.acquire().await.unwrap();
                // Try to acquire the first semaphore - this can deadlock
                // if task 1 is already holding semaphore1
                let permit1 = semaphore1.acquire().await.unwrap();
                // Release both semaphores
                drop(permit1);
                drop(permit2);
            }
            .primary()
            .spawn_named("Task 2");
        }
    })();
}
