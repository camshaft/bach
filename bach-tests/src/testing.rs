use std::{collections::HashMap, sync::Mutex};

pub fn init_tracing() {
    if !cfg!(test) || cfg!(feature = "leaks") {
        return;
    }

    use std::sync::Once;

    static TRACING: Once = Once::new();

    // make sure this only gets initialized once
    TRACING.call_once(|| {
        let format = tracing_subscriber::fmt::format()
            .with_level(false) // don't include levels in formatted output
            .with_timer(Uptime)
            .with_ansi(false)
            .compact(); // Use a less verbose output format.

        struct Uptime;

        // Generate the timestamp from the testing IO provider rather than wall clock.
        impl tracing_subscriber::fmt::time::FormatTime for Uptime {
            fn format_time(
                &self,
                w: &mut tracing_subscriber::fmt::format::Writer<'_>,
            ) -> std::fmt::Result {
                let now =
                    bach::time::scheduler::scope::try_borrow_mut_with(|s| Some(s.as_ref()?.now()));
                if let Some(now) = now {
                    write!(w, "{now}")
                } else {
                    write!(w, "[UNKNOWN]")
                }
            }
        }

        let env_filter = tracing_subscriber::EnvFilter::builder()
            .with_default_directive(tracing::Level::DEBUG.into())
            .with_env_var("BACH_LOG")
            .from_env()
            .unwrap();

        tracing_subscriber::fmt()
            .with_env_filter(env_filter)
            .event_format(format)
            .with_test_writer()
            .init();
    });
}

pub trait Event: Clone + core::fmt::Debug + Eq + core::hash::Hash {
    fn is_start(&self) -> bool;
}

pub struct Log<T>(Mutex<Vec<T>>);

impl<T: Event> Default for Log<T> {
    fn default() -> Self {
        Self::new()
    }
}

impl<T: Event> Log<T> {
    pub const fn new() -> Self {
        Self(Mutex::new(Vec::new()))
    }

    pub fn push(&self, event: T) {
        self.0.lock().unwrap().push(event);
    }

    pub fn check(&self) -> Vec<T> {
        let mut sequences = vec![];
        let mut current_sequence = vec![];
        let mut seen = HashMap::<_, usize>::new();

        let events = core::mem::take(&mut *self.0.lock().unwrap());

        for event in events.iter() {
            if event.is_start() {
                if !current_sequence.is_empty() {
                    let seq = std::mem::take(&mut current_sequence);
                    *seen.entry(seq.clone()).or_default() += 1;
                    sequences.push(seq);
                }
            } else {
                current_sequence.push(event.clone());
            }
        }

        if !current_sequence.is_empty() {
            let seq = std::mem::take(&mut current_sequence);
            *seen.entry(seq.clone()).or_default() += 1;
            sequences.push(seq);
        }

        let mut duplicate = false;

        for (seq, count) in seen {
            if count == 1 {
                continue;
            }
            duplicate = true;
            eprintln!("duplicate sequence ({count}): {seq:#?}");
        }

        assert!(!duplicate, "duplicate interleavings found");
        assert!(!sequences.is_empty(), "no event sequences recorded");

        eprintln!("Found {} unique event sequences", sequences.len());

        events
    }
}
