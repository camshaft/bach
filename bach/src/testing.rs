pub fn init_tracing() {
    #[cfg(feature = "tracing")]
    init_tracing_impl();
}

#[cfg(feature = "tracing")]
fn init_tracing_impl() {
    use std::sync::Once;

    static TRACING: Once = Once::new();

    // make sure this only gets initialized once
    TRACING.call_once(|| {
        let format = tracing_subscriber::fmt::format()
            .with_level(false) // don't include levels in formatted output
            .with_timer(Uptime)
            .with_ansi(false)
            .compact(); // Use a less verbose output format.

        struct Uptime;

        // Generate the timestamp from the testing IO provider rather than wall clock.
        impl tracing_subscriber::fmt::time::FormatTime for Uptime {
            fn format_time(
                &self,
                w: &mut tracing_subscriber::fmt::format::Writer<'_>,
            ) -> std::fmt::Result {
                let now =
                    crate::time::scheduler::scope::try_borrow_mut_with(|s| Some(s.as_ref()?.now()));
                if let Some(now) = now {
                    write!(w, "{now}")
                } else {
                    write!(w, "[UNKNOWN]")
                }
            }
        }

        let env_filter = tracing_subscriber::EnvFilter::builder()
            .with_default_directive(tracing::Level::DEBUG.into())
            .with_env_var("BACH_LOG")
            .from_env()
            .unwrap();

        tracing_subscriber::fmt()
            .with_env_filter(env_filter)
            .event_format(format)
            .with_test_writer()
            .init();
    });
}
