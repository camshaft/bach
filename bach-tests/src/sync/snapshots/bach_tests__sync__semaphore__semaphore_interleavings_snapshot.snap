---
source: bach-tests/src/sync/semaphore.rs
expression: LOG.check()
---
[
    Start,
    SemaphoreAcquired {
        task: 0,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 1,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 2,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 3,
        permits: 2,
    },
    SemaphoreReleased {
        task: 0,
        permits: 1,
    },
    SemaphoreReleased {
        task: 1,
        permits: 2,
    },
    SemaphoreReleased {
        task: 2,
        permits: 1,
    },
    SemaphoreReleased {
        task: 3,
        permits: 2,
    },
    Start,
    SemaphoreAcquired {
        task: 0,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 1,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 3,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 2,
        permits: 1,
    },
    SemaphoreReleased {
        task: 0,
        permits: 1,
    },
    SemaphoreReleased {
        task: 1,
        permits: 2,
    },
    SemaphoreReleased {
        task: 3,
        permits: 2,
    },
    SemaphoreReleased {
        task: 2,
        permits: 1,
    },
    Start,
    SemaphoreAcquired {
        task: 0,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 2,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 1,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 3,
        permits: 2,
    },
    SemaphoreReleased {
        task: 0,
        permits: 1,
    },
    SemaphoreReleased {
        task: 2,
        permits: 1,
    },
    SemaphoreReleased {
        task: 1,
        permits: 2,
    },
    SemaphoreReleased {
        task: 3,
        permits: 2,
    },
    Start,
    SemaphoreAcquired {
        task: 0,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 2,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 3,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 1,
        permits: 2,
    },
    SemaphoreReleased {
        task: 0,
        permits: 1,
    },
    SemaphoreReleased {
        task: 2,
        permits: 1,
    },
    SemaphoreReleased {
        task: 3,
        permits: 2,
    },
    SemaphoreReleased {
        task: 1,
        permits: 2,
    },
    Start,
    SemaphoreAcquired {
        task: 0,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 3,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 2,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 1,
        permits: 2,
    },
    SemaphoreReleased {
        task: 0,
        permits: 1,
    },
    SemaphoreReleased {
        task: 3,
        permits: 2,
    },
    SemaphoreReleased {
        task: 2,
        permits: 1,
    },
    SemaphoreReleased {
        task: 1,
        permits: 2,
    },
    Start,
    SemaphoreAcquired {
        task: 0,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 3,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 1,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 2,
        permits: 1,
    },
    SemaphoreReleased {
        task: 0,
        permits: 1,
    },
    SemaphoreReleased {
        task: 3,
        permits: 2,
    },
    SemaphoreReleased {
        task: 1,
        permits: 2,
    },
    SemaphoreReleased {
        task: 2,
        permits: 1,
    },
    Start,
    SemaphoreAcquired {
        task: 1,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 0,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 2,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 3,
        permits: 2,
    },
    SemaphoreReleased {
        task: 1,
        permits: 2,
    },
    SemaphoreReleased {
        task: 0,
        permits: 1,
    },
    SemaphoreReleased {
        task: 2,
        permits: 1,
    },
    SemaphoreReleased {
        task: 3,
        permits: 2,
    },
    Start,
    SemaphoreAcquired {
        task: 1,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 0,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 3,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 2,
        permits: 1,
    },
    SemaphoreReleased {
        task: 1,
        permits: 2,
    },
    SemaphoreReleased {
        task: 0,
        permits: 1,
    },
    SemaphoreReleased {
        task: 3,
        permits: 2,
    },
    SemaphoreReleased {
        task: 2,
        permits: 1,
    },
    Start,
    SemaphoreAcquired {
        task: 1,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 2,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 0,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 3,
        permits: 2,
    },
    SemaphoreReleased {
        task: 1,
        permits: 2,
    },
    SemaphoreReleased {
        task: 2,
        permits: 1,
    },
    SemaphoreReleased {
        task: 0,
        permits: 1,
    },
    SemaphoreReleased {
        task: 3,
        permits: 2,
    },
    Start,
    SemaphoreAcquired {
        task: 1,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 2,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 3,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 0,
        permits: 1,
    },
    SemaphoreReleased {
        task: 1,
        permits: 2,
    },
    SemaphoreReleased {
        task: 2,
        permits: 1,
    },
    SemaphoreReleased {
        task: 3,
        permits: 2,
    },
    SemaphoreReleased {
        task: 0,
        permits: 1,
    },
    Start,
    SemaphoreAcquired {
        task: 1,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 3,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 2,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 0,
        permits: 1,
    },
    SemaphoreReleased {
        task: 1,
        permits: 2,
    },
    SemaphoreReleased {
        task: 3,
        permits: 2,
    },
    SemaphoreReleased {
        task: 2,
        permits: 1,
    },
    SemaphoreReleased {
        task: 0,
        permits: 1,
    },
    Start,
    SemaphoreAcquired {
        task: 1,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 3,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 0,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 2,
        permits: 1,
    },
    SemaphoreReleased {
        task: 1,
        permits: 2,
    },
    SemaphoreReleased {
        task: 3,
        permits: 2,
    },
    SemaphoreReleased {
        task: 0,
        permits: 1,
    },
    SemaphoreReleased {
        task: 2,
        permits: 1,
    },
    Start,
    SemaphoreAcquired {
        task: 2,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 1,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 0,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 3,
        permits: 2,
    },
    SemaphoreReleased {
        task: 2,
        permits: 1,
    },
    SemaphoreReleased {
        task: 1,
        permits: 2,
    },
    SemaphoreReleased {
        task: 0,
        permits: 1,
    },
    SemaphoreReleased {
        task: 3,
        permits: 2,
    },
    Start,
    SemaphoreAcquired {
        task: 2,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 1,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 3,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 0,
        permits: 1,
    },
    SemaphoreReleased {
        task: 2,
        permits: 1,
    },
    SemaphoreReleased {
        task: 1,
        permits: 2,
    },
    SemaphoreReleased {
        task: 3,
        permits: 2,
    },
    SemaphoreReleased {
        task: 0,
        permits: 1,
    },
    Start,
    SemaphoreAcquired {
        task: 2,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 0,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 1,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 3,
        permits: 2,
    },
    SemaphoreReleased {
        task: 2,
        permits: 1,
    },
    SemaphoreReleased {
        task: 0,
        permits: 1,
    },
    SemaphoreReleased {
        task: 1,
        permits: 2,
    },
    SemaphoreReleased {
        task: 3,
        permits: 2,
    },
    Start,
    SemaphoreAcquired {
        task: 2,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 0,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 3,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 1,
        permits: 2,
    },
    SemaphoreReleased {
        task: 2,
        permits: 1,
    },
    SemaphoreReleased {
        task: 0,
        permits: 1,
    },
    SemaphoreReleased {
        task: 3,
        permits: 2,
    },
    SemaphoreReleased {
        task: 1,
        permits: 2,
    },
    Start,
    SemaphoreAcquired {
        task: 2,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 3,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 0,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 1,
        permits: 2,
    },
    SemaphoreReleased {
        task: 2,
        permits: 1,
    },
    SemaphoreReleased {
        task: 3,
        permits: 2,
    },
    SemaphoreReleased {
        task: 0,
        permits: 1,
    },
    SemaphoreReleased {
        task: 1,
        permits: 2,
    },
    Start,
    SemaphoreAcquired {
        task: 2,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 3,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 1,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 0,
        permits: 1,
    },
    SemaphoreReleased {
        task: 2,
        permits: 1,
    },
    SemaphoreReleased {
        task: 3,
        permits: 2,
    },
    SemaphoreReleased {
        task: 1,
        permits: 2,
    },
    SemaphoreReleased {
        task: 0,
        permits: 1,
    },
    Start,
    SemaphoreAcquired {
        task: 3,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 1,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 2,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 0,
        permits: 1,
    },
    SemaphoreReleased {
        task: 3,
        permits: 2,
    },
    SemaphoreReleased {
        task: 1,
        permits: 2,
    },
    SemaphoreReleased {
        task: 2,
        permits: 1,
    },
    SemaphoreReleased {
        task: 0,
        permits: 1,
    },
    Start,
    SemaphoreAcquired {
        task: 3,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 1,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 0,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 2,
        permits: 1,
    },
    SemaphoreReleased {
        task: 3,
        permits: 2,
    },
    SemaphoreReleased {
        task: 1,
        permits: 2,
    },
    SemaphoreReleased {
        task: 0,
        permits: 1,
    },
    SemaphoreReleased {
        task: 2,
        permits: 1,
    },
    Start,
    SemaphoreAcquired {
        task: 3,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 2,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 1,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 0,
        permits: 1,
    },
    SemaphoreReleased {
        task: 3,
        permits: 2,
    },
    SemaphoreReleased {
        task: 2,
        permits: 1,
    },
    SemaphoreReleased {
        task: 1,
        permits: 2,
    },
    SemaphoreReleased {
        task: 0,
        permits: 1,
    },
    Start,
    SemaphoreAcquired {
        task: 3,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 2,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 0,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 1,
        permits: 2,
    },
    SemaphoreReleased {
        task: 3,
        permits: 2,
    },
    SemaphoreReleased {
        task: 2,
        permits: 1,
    },
    SemaphoreReleased {
        task: 0,
        permits: 1,
    },
    SemaphoreReleased {
        task: 1,
        permits: 2,
    },
    Start,
    SemaphoreAcquired {
        task: 3,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 0,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 2,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 1,
        permits: 2,
    },
    SemaphoreReleased {
        task: 3,
        permits: 2,
    },
    SemaphoreReleased {
        task: 0,
        permits: 1,
    },
    SemaphoreReleased {
        task: 2,
        permits: 1,
    },
    SemaphoreReleased {
        task: 1,
        permits: 2,
    },
    Start,
    SemaphoreAcquired {
        task: 3,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 0,
        permits: 1,
    },
    SemaphoreAcquired {
        task: 1,
        permits: 2,
    },
    SemaphoreAcquired {
        task: 2,
        permits: 1,
    },
    SemaphoreReleased {
        task: 3,
        permits: 2,
    },
    SemaphoreReleased {
        task: 0,
        permits: 1,
    },
    SemaphoreReleased {
        task: 1,
        permits: 2,
    },
    SemaphoreReleased {
        task: 2,
        permits: 1,
    },
]
